src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs

```diff
@@ -0,0 +1,399 @@
+            private readonly IScheduler _scheduler;
+            private readonly ISourceCache<TObject, TKey> _source;
+            private readonly IDisposable _sourceSubscription;
+            private readonly Func<TObject, TimeSpan?> _timeSelector;
+
+            private bool _hasSourceCompleted;
+            private ScheduledManagement? _nextScheduledManagement;
+
+            protected SubscriptionBase(
+                IObserver<IEnumerable<KeyValuePair<TKey, TObject>>> observer,
+                IScheduler? scheduler,
+                ISourceCache<TObject, TKey> source,
+                Func<TObject, TimeSpan?> timeSelector)
+            {
+                _observer = observer;
+                _source = source;
+                _timeSelector = timeSelector;
+
+                _scheduler = scheduler ?? GlobalConfig.DefaultScheduler;
+
+                _expirationDueTimesByKey = new();
+                _proposedExpirationsQueue = new();
+                _removedItemsBuffer = new();
+
+                _sourceSubscription = source
+                    .Connect()
+                    // It's important to set this flag outside the context of a lock, because it'll be read outside of lock as well.
+                    .Finally(() => _hasSourceCompleted = true)
+                    .Synchronize(SynchronizationGate)
+                    .SubscribeSafe(Observer.Create<IChangeSet<TObject, TKey>>(
+                        onNext: OnSourceNext,
+                        onError: OnSourceError,
+                        onCompleted: OnSourceCompleted));
+            }
+
+            public void Dispose()
+            {
+                lock (SynchronizationGate)
+                {
+                    _sourceSubscription.Dispose();
+
+                    TryCancelNextScheduledManagement();
+                }
+            }
+
+            protected IScheduler Scheduler
+                => _scheduler;
+
+            // Instead of using a dedicated _synchronizationGate object, we can save an allocation by using any object that is never exposed to public consumers.
+            protected object SynchronizationGate
+                => _expirationDueTimesByKey;
+
+            protected abstract DateTimeOffset? GetNextManagementDueTime();
+
+            protected DateTimeOffset? GetNextProposedExpirationDueTime()
+                => _proposedExpirationsQueue.Count is 0
+                    ? null
+                    : _proposedExpirationsQueue[0].DueTime;
+
+            protected abstract void OnExpirationsManaged(DateTimeOffset dueTime);
+
+            private void ClearExpiration(TKey key)
+                // This is what puts the "proposed" in _proposedExpirationsQueue.
+                // Finding the position of the item to remove from the queue would be O(log n), at best,
+                // so just leave it and flush it later during normal processing of the queue.
+                => _expirationDueTimesByKey.Remove(key);
```

Comment:
So, to avoid a linear search, you just leave the expire times in the Queue and later check for invalid ones?  That's good.  I like that.

Reply:
Yup. Works well for caches, since we can do a key lookup, but it was a nightmare to implement for lists, and it ended up being costlier for performance, anyway.